src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/CertificateData.cs

```diff
@@ -59,126 +61,57 @@ internal CertificateData(byte[] rawData)
         try
         {
 #endif
-            DerSequenceReader reader = new DerSequenceReader(rawData);
-
-            DerSequenceReader tbsCertificate = reader.ReadSequence();
-
-            if (tbsCertificate.PeekTag() == DerSequenceReader.ContextSpecificConstructedTag0)
-            {
-                DerSequenceReader version = tbsCertificate.ReadSequence();
-                Version = version.ReadInteger();
-            }
-            else if (tbsCertificate.PeekTag() != (byte)DerSequenceReader.DerTag.Integer)
-            {
-                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
-            }
-            else
-            {
-                Version = 0;
-            }
+            CertificateAsn certificate = AsnSerializer.Deserialize<CertificateAsn>(rawData, AsnEncodingRules.DER);
+            TbsCertificateAsn tbsCertificate = certificate.TbsCertificate;
 
+            Version = tbsCertificate.Version;
             if (Version < 0 || Version > 2)
                 throw new CryptographicException();
 
-            SerialNumber = tbsCertificate.ReadIntegerBytes();
+            SerialNumber = tbsCertificate.SerialNumber.ToArray();
```

Comment:
Why? Just store the TbsCertificateAsn and do the ToArray when asked.

Reply:
Actually the whole `CertificateData` class could basically be replaced with `CertificateAsn` and few added caches for Issuer and Subject stored as `X500DistinguishedName`. I kept it around only temporarily.